lib/operations/crud_write.js

```diff
@@ -0,0 +1,284 @@
+const applyWriteConcern = require('../utils').applyWriteConcern;
+const handleCallback = require('../utils').handleCallback;
+const decorateWithCollation = require('../utils').decorateWithCollation;
+const toError = require('../utils').toError;
+
+const INSERT = 'insert';
+const UPDATE = 'update';
+const REMOVE = 'remove';
+
+const RETRYABLE_WIRE_VERSION = 6;
+
+function isSingleServer(topology) {
+  if (topology.type && topology.type === 'server') {
+    return true;
+  }
+
+  if (topology.description && topology.description.type === 'Single') {
+    return true;
+  }
+
+  return false;
+}
+
+function isRetryabilitySupported(topology) {
+  if (isSingleServer(topology)) {
+    return false;
+  }
+
+  const maxWireVersion = topology.lastIsMaster().maxWireVersion;
+  if (maxWireVersion < RETRYABLE_WIRE_VERSION) {
+    return false;
+  }
+
+  if (!topology.logicalSessionTimeoutMinutes) {
+    return false;
+  }
+
+  return true;
+}
+
+class CrudWriteOperation extends OperationBase {
+  constructor(model) {
+    super(model.options);
+    this.model = model;
+  }
+
+  canRetry(server, topology) {
+    return (
+      this.options.retryWrites &&
+      this.session &&
+      !this.session.inTransaction() &&
+      isRetryabilitySupported(topology) &&
+      this.model.canRetry()
+    );
+  }
+
+  enableRetry() {
+    this.model.options.willRetryWrite = true;
+    this.session.incrementTransactionNumber();
+  }
+
+  execute(server, callback) {
+    return this.model.execute(server, callback);
+  }
+}
+
+class CrudModel {
+  constructor(collection, options) {
+    options = applyRetryableWrites(options, collection.s.db);
+    options = applyWriteConcern(options, { db: collection.s.db, collection }, options);
+    options.serializeFunctions = options.serializeFunctions || collection.s.serializeFunctions;
+    if (collection.s.options.ignoreUndefined) {
+      options.ignoreUndefined = collection.s.options.ignoreUndefined;
+    }
+
+    this.options = options;
+    this.namespace = collection.s.namespace;
+  }
+
+  execute(server, callback) {
+    if (!this.type || !this.operations) {
+      // TODO: Throw something?
+    }
+    server[this.type](this.namespace.toString(), this.operations, this.options, (err, result) => {
+      if (callback == null) return;
+      if (err) return handleCallback(callback, err, null);
+      if (result == null) return handleCallback(callback, null, null);
+      if (result.result.code) return handleCallback(callback, toError(result.result));
+      if (result.result.writeErrors)
+        return handleCallback(callback, toError(result.result.writeErrors[0]));
+      // Return the results
+      handleCallback(callback, null, result);
+    });
+  }
+}
+
+defineAspects(CrudWriteOperation, [Aspect.EXECUTE_WITH_SELECTION, Aspect.RETRY]);
+
+class BulkInsertModel extends CrudModel {
+  constructor(collection, options) {
+    super(collection, options);
+    this.type = INSERT;
```

Comment:
why not pass `type` in the constructor?

Reply:
IMO that would make the constructor more complicated and less readable.

Reply:
Can you refine what you mean by "complicated" and "less readable"?

I find it far less readable that each of the subclasses of the `CrudModel` attach a property called `type` onto the instance. Doing something like `super(INSERT, collection, options)` shows me very clearly that this is an `INSERT` CrudModel.

Another thing to think about here is that this `type` thing is _very_ tailored to the existing `Server` implementation, where you are calling some method on the `Server` class that corresponds with the type. I understand that we need to move piecemeal in a direction of just using the `command` method, but I'm concerned that baking in assumptions like this could be prolonging the issue.

One way to approach this differently would be to make the models basic read-only data types encompassing the operation you wish to perform, and switching on model type in the `execute` method of a meta bulk operation. 

```
class BulkWriteOperation {
   constructor(models) {
      this.models = models;
   }

   execute(server, callback) {
     this.models.forEach(model => {
       const writeMethod = resolveWriteMethod(model);
       server[writeMethod](...., (err, result) => {
          // do required synchronization to make sure callback is called after all models
       });
     });
   }
}
```

This way we could degrade gracefully when we remove those methods and only provide a single `command` method on the server instance.

Reply:
As you pointed out, `type` is going to go away pretty soon, once we switch to all of them using `command`. I'd rather have it just be a property we can change so that when we do remove it, we are not messing with the constructor signature. A `resolveWriteMethod` seems like a bit of overkill in this case.